File: frontend/src/types/strategy.ts
/**
 * Types for Backend Strategy API
 * Version: 4.0 - Clean Structure with Multiple Strategy Types
 * Date: 02/03/2026
 *
 * Type 1: Allocation by Weight ✅
 * Type 2: Health Factor Target ✅
 * Type 3: Liquidity Range Monitor (Future)
 */

/**
 * Strategy Type enum matching backend implementation
 */
export enum StrategyType {
  AllocationByWeight = 1, // ✅ Implemented
  HealthFactorTarget = 2, // ✅ Implemented
  LiquidityRangeMonitor = 3, // 🔜 Future
  YieldThreshold = 4, // 🔜 Future
  ProtocolDiversification = 5, // 🔜 Future
  ChainAllocation = 6, // 🔜 Future
  AssetTypeAllocation = 7, // 🔜 Future
  BestPurchaseWindow = 8, // ✅ Implemented
}

export type WindowPeriod = 'h1' | 'h24' | 'd7' | 'd14' | 'd30' | 'd90' | 'd200' | 'y1' | 'ytd';
export type WindowDirection = 'min' | 'max';
export type PriceDirection = 'above' | 'below';

export type PurchaseTrigger =
  | { type: 'window'; window: WindowPeriod; direction: WindowDirection }
  | { type: 'price'; target: number; direction: PriceDirection };

export interface BestPurchaseWindowEntry {
  assetKey: string;
  symbol: string;
  coingeckoId: string;
  protocol: ProtocolInfo;
  chain: ChainInfo;
  token: TokenInfo;
  /** @deprecated use `triggers` */
  trigger?: PurchaseTrigger;
  triggers?: PurchaseTrigger[];
}

export interface TriggerEvaluation {
  trigger: PurchaseTrigger;
  referencePrice: number;
  referenceLabel: string;
  signal: boolean;
}

export interface TokenMarketData {
  image?: string;
  currentPrice?: number;
  lastUpdated?: string;
  marketCap?: number;
  marketCapRank?: number;
  fullyDilutedValuation?: number;
  totalVolume?: number;
  high24h?: number;
  low24h?: number;
  priceChange24h?: number;
  priceChangePercentage1h?: number;
  priceChangePercentage24h?: number;
  priceChangePercentage7d?: number;
  priceChangePercentage14d?: number;
  priceChangePercentage30d?: number;
  priceChangePercentage200d?: number;
  priceChangePercentage1y?: number;
  priceChangePercentageYtd?: number;
  marketCapChange24h?: number;
  marketCapChangePercentage24h?: number;
  circulatingSupply?: number;
  totalSupply?: number;
  maxSupply?: number;
  ath?: number;
  athChangePercentage?: number;
  athDate?: string;
  atl?: number;
  atlChangePercentage?: number;
  atlDate?: string;
}

export interface PurchaseWindowEvalResult {
  assetKey: string;
  symbol: string;
  currentPriceUsd: number;
  /** @deprecated use `evaluations` */
  trigger?: PurchaseTrigger;
  signal: boolean;
  referencePrice: number;
  referenceLabel: string;
  evaluations?: TriggerEvaluation[];
  fetchedAt: string;
  marketData?: TokenMarketData;
  dataUnavailable?: boolean;
  errorMessage?: string;
}

/**
 * Protocol information in strategy
 */
export interface ProtocolInfo {
  id: string;
  name: string;
}

/**
 * Chain information in strategy
 */
export interface ChainInfo {
  id: string;
  name: string;
}

/**
 * Token information in strategy
 */
export interface TokenInfo {
  symbol: string;
  name: string;
  address: string;
  logo: string | null;
}

/**
 * Allocation item in Type 1 strategy
 */
export interface AllocationItem {
  assetKey: string;
  protocol: ProtocolInfo;
  chain: ChainInfo;
  token: TokenInfo;
  group: string; // "Lending Supply", "Lending Borrow", etc.
  groupType: number; // 10, 11, etc.
  targetWeight: number; // 0-100 (STATIC: user defined)
  positionType: number; // 3 = Supplied, 4 = Borrowed
  displayOrder?: number; // Display order in UI
}

/**
 * Health Factor target in Type 2 strategy
 */
export interface HealthFactorTarget {
  assetKey: string;
  protocol: ProtocolInfo;
  chain: ChainInfo;
  targetHealthFactor: number; // STATIC: user defined (e.g., 2.5)
  criticalThreshold: number; // STATIC: user defined (e.g., 1.5)
  displayOrder?: number; // Display order in UI
}

/**
 * Base strategy interface (common fields)
 */
export interface BaseStrategy {
  id: string;
  walletGroupId: string;
  strategyType: number;
  name: string;
  description?: string | null;
  displayOrder?: number; // Display order in UI
  createdAt: string;
  updatedAt: string;
}

/**
 * Type 1: Allocation by Weight Strategy
 */
export interface AllocationStrategy extends BaseStrategy {
  strategyType: 1;
  allocations: AllocationItem[];
}

/**
 * Type 2: Health Factor Target Strategy
 */
export interface HealthFactorStrategy extends BaseStrategy {
  strategyType: 2;
  targets: HealthFactorTarget[];
}

/**
 * Type 8: Best Purchase Window Strategy
 */
export interface BestPurchaseWindowStrategy extends BaseStrategy {
  strategyType: 8;
  purchaseWindowEntries: BestPurchaseWindowEntry[];
  purchaseWindowResults?: PurchaseWindowEvalResult[];
}

/**
 * Union type for all strategy types
 */
export type Strategy = AllocationStrategy | HealthFactorStrategy | BestPurchaseWindowStrategy;

/**
 * Response from GET /api/strategies/:walletGroupId
 */
export interface GetStrategiesResponse {
  walletGroupId: string;
  strategies: Strategy[];
}

/**
 * @deprecated Use SaveAllocationStrategyRequest | SaveHealthFactorStrategyRequest instead
 */
export interface LegacySaveStrategyRequest {
  walletGroupId: string;
  strategyId?: string;
  strategyType: number;
  name?: string | null;
  description?: string | null;
  allocations: Array<{
    assetKey: string;
    protocol: string;
    chain: string;
    group: string;
    targetWeight: number;
  }>;
}

/**
 * Response from creating/updating a strategy
 */
export interface SaveStrategyResponse {
  key: string;
  itemsCount: number;
  wallets: string[];
  savedAt: string;
}

/**
 * Allocation delta calculation result
 * Shows the difference between target and current allocation
 */
export interface AllocationDelta {
  assetKey: string;
  group: string;
  targetWeight: number;
  currentWeight: number;
  deltaWeight: number; // currentWeight - targetWeight
  targetValueUsd: number;
  currentValueUsd: number;
  deltaValueUsd: number; // currentValueUsd - targetValueUsd
  needsRebalance: boolean; // true if |deltaWeight| > threshold (e.g., 5%)
}

/**
 * Group normalization mapping
 */
export const GroupTypeMapping = {
  Lending: 'LendingAndBorrowing',
  Liquidity: 'LiquidityPool',
  Staking: 'Staking',
  Wallet: 'Wallet',
  All: null,
} as const;

export type GroupType = keyof typeof GroupTypeMapping;

// ============================================
// TYPE 2: HEALTH FACTOR TARGET
// ============================================

/**
 * Configuration for Health Factor Target strategy (Type 2)
 */
export interface HealthFactorTargetConfig {
  targetHealthFactor: number; // Desired HF (e.g., 2.0)
  warningThreshold: number; // Warning level (e.g., 1.8)
  criticalThreshold: number; // Critical level (e.g., 1.5)
  autoSuggest: boolean; // Generate action suggestions
  protocols: string[]; // Protocol IDs to monitor
}

/**
 * Action suggestion for health factor management
 */
export interface HealthFactorAction {
  action: 'add_collateral' | 'reduce_debt' | 'close_position';
  protocol: string;
  assetKey: string;
  amountUsd: number;
  priority: 'critical' | 'high' | 'medium' | 'low';
}

/**
 * Current health factor status
 */
export interface HealthFactorStatus {
  current: number; // Current HF value
  target: number; // Target HF value
  criticalThreshold?: number; // Critical threshold (alert level) for this position
  status: 'safe' | 'warning' | 'critical' | 'danger';
  needsAction: boolean;
  collateralValue: number; // Total collateral in USD
  debtValue: number; // Total debt in USD
  totalValue: number; // Total value (collateral + debt) in USD
  suggestions: HealthFactorAction[];
  protocol: string; // Protocol ID
  protocolName: string; // Protocol display name
  protocolLogo?: string; // Protocol logo URL
  chain: string; // Chain name
  chainLogo?: string; // Chain logo URL
}

/**save Type 1: Allocation Strategy
 */
export interface SaveAllocationStrategyRequest {
  id?: string; // For updates (if present, backend updates; if absent, creates)
  strategyType: 1;
  name: string;
  description?: string | null;
  allocations: Array<{
    assetKey: string;
    protocol: {
      id: string;
      name: string;
      logo: string;
    };
    chain: {
      id: string;
      name: string;
      logo: string;
    };
    token: {
      symbol: string;
      name: string;
      address: string;
      logo: string;
    };
    group: string; // "Lending Supply"
    groupType: number; // 10 = Lending, 20 = Liquidity, etc.
    targetWeight: number;
    positionType: number; // 0 = Default, 1 = Supplied, 2 = Borrowed
  }>;
}

/**
 * Request to save Type 2: Health Factor Strategy
 */
export interface SaveHealthFactorStrategyRequest {
  walletGroupId: string;
  strategyId?: string;
  strategyType: 2;
  name: string;
  description?: string | null;
  targets: Array<{
    assetKey: string;
    protocol: string;
    chain: string;
    targetHealthFactor: number;
    criticalThreshold: number;
  }>;
}

export interface SaveBestPurchaseWindowStrategyRequest {
  id?: string;
  strategyType: 8;
  name: string;
  description?: string | null;
  purchaseWindowEntries: BestPurchaseWindowEntry[];
}

/**
 * Union type for save requests
 */
export type SaveStrategyRequest =
  | SaveAllocationStrategyRequest
  | SaveHealthFactorStrategyRequest
  | SaveBestPurchaseWindowStrategyRequest;

/**
 * Legacy type alias for target allocation
 */
export interface TargetAllocation {
  assetKey: string;
  group: string;
  targetWeight: number;
  protocol?: string;
  chain?: string;
}

/**
 * Legacy type alias for strategy item
 */
export type StrategyItem = AllocationItem | HealthFactorTarget;

/**
 * Token metadata for strategy assets
 */
export interface TokenMetadata {
  type: number | null;
  symbol: string;
  name: string;
  address: string | null;
  logo: string | null;
  chain: string;
}

/**
 * Protocol metadata for strategy assets
 */
export interface ProtocolMetadata {
  id: string;
  name: string;
  chain: string;
  url: string | null;
  logo: string | null;
}

/**
 * Complete metadata for a strategy asset
 */
export interface StrategyAssetMetadata {
  symbol: string | null;
  name: string | null;
  address: string | null;
  chainId: string | null;
  chain: string | null;
  protocol: ProtocolMetadata | null;
  positionLabel: string | null;
  positionType: number;
  tokens: TokenMetadata[];
}

/**
 * Legacy type alias for strategy asset
 */
export type StrategyAsset = AllocationItem;

/**
 * Strategy data in multi-strategy response
 */
export interface StrategyData {
  id: string;
  strategyType: number;
  name: string;
  description?: string | null;
  allocations?: AllocationItem[];
  targets?: HealthFactorTarget[];
  purchaseWindowEntries?: BestPurchaseWindowEntry[];
  purchaseWindowResults?: PurchaseWindowEvalResult[];
  createdAt?: string;
  updatedAt?: string;
}

/**
 * Request to save multiple strategies
 */
export interface SaveStrategiesRequest {
  walletGroupId: string;
  strategies: Array<{
    id?: string;
    strategyType: number;
    name: string;
    description?: string | null;
    allocations?: unknown[];
    targets?: unknown[];
    purchaseWindowEntries?: BestPurchaseWindowEntry[];
  }>;
}

/**
 * Response from GET/POST multi-strategy endpoint
 */
export interface SaveStrategiesResponse {
  walletGroupId: string;
  strategies: StrategyData[];
  wallets: string[];
  count: number;
}

File: backend-rust/crates/core/src/strategy.rs
use chrono::{DateTime, Datelike, TimeZone, Utc};
use serde::{Deserialize, Serialize};
use serde_repr::{Deserialize_repr, Serialize_repr};
use uuid::Uuid;

fn serialize_uuid<S>(uuid: &Uuid, serializer: S) -> Result<S::Ok, S::Error>
where
    S: serde::Serializer,
{
    serializer.serialize_str(&uuid.to_string())
}

fn deserialize_uuid<'de, D>(deserializer: D) -> Result<Uuid, D::Error>
where
    D: serde::Deserializer<'de>,
{
    let s = String::deserialize(deserializer)?;
    Ok(Uuid::parse_str(&s).unwrap_or(Uuid::nil()))
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize_repr, Deserialize_repr)]
#[repr(i32)]
pub enum StrategyType {
    AllocationByWeight = 1,
    HealthFactorTarget = 2,
    BestPurchaseWindow = 8,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WindowPeriod {
    H1,
    H24,
    D7,
    D14,
    D30,
    D90,
    D200,
    Y1,
    Ytd,
}

impl WindowPeriod {
    pub fn start_unix(&self, now: i64) -> i64 {
        match self {
            WindowPeriod::H1 => now - 3_600,
            WindowPeriod::H24 => now - 86_400,
            WindowPeriod::D7 => now - 7 * 86_400,
            WindowPeriod::D14 => now - 14 * 86_400,
            WindowPeriod::D30 => now - 30 * 86_400,
            WindowPeriod::D90 => now - 90 * 86_400,
            WindowPeriod::D200 => now - 200 * 86_400,
            WindowPeriod::Y1 => now - 365 * 86_400,
            WindowPeriod::Ytd => {
                let dt = Utc.timestamp_opt(now, 0).single().unwrap_or_else(Utc::now);
                Utc.with_ymd_and_hms(dt.year(), 1, 1, 0, 0, 0)
                    .single()
                    .map(|d| d.timestamp())
                    .unwrap_or(now)
            }
        }
    }

    pub fn label(&self) -> &'static str {
        match self {
            WindowPeriod::H1 => "1h",
            WindowPeriod::H24 => "24h",
            WindowPeriod::D7 => "7 dias",
            WindowPeriod::D14 => "14 dias",
            WindowPeriod::D30 => "30 dias",
            WindowPeriod::D90 => "90 dias",
            WindowPeriod::D200 => "200 dias",
            WindowPeriod::Y1 => "1 ano",
            WindowPeriod::Ytd => "ano atual",
        }
    }

    pub fn order(&self) -> u8 {
        match self {
            WindowPeriod::H1 => 0,
            WindowPeriod::H24 => 1,
            WindowPeriod::D7 => 2,
            WindowPeriod::D14 => 3,
            WindowPeriod::D30 => 4,
            WindowPeriod::D90 => 5,
            WindowPeriod::D200 => 6,
            WindowPeriod::Y1 => 7,
            WindowPeriod::Ytd => 8,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WindowDirection {
    Min,
    Max,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PriceDirection {
    Above,
    Below,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case", tag = "type")]
pub enum PurchaseTrigger {
    Window {
        window: WindowPeriod,
        direction: WindowDirection,
    },
    Price {
        target: f64,
        direction: PriceDirection,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BestPurchaseWindowEntry {
    pub asset_key: String,
    pub symbol: String,
    pub coingecko_id: String,
    pub protocol: AllocationProtocol,
    pub chain: AllocationChain,
    pub token: AllocationToken,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub trigger: Option<PurchaseTrigger>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub triggers: Vec<PurchaseTrigger>,
}

impl BestPurchaseWindowEntry {
    pub fn effective_triggers(&self) -> Vec<PurchaseTrigger> {
        if !self.triggers.is_empty() {
            self.triggers.clone()
        } else if let Some(t) = &self.trigger {
            vec![t.clone()]
        } else {
            Vec::new()
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct TokenMarketData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub image: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub current_price: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_updated: Option<DateTime<Utc>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub market_cap: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub market_cap_rank: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fully_diluted_valuation: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub total_volume: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub high_24h: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub low_24h: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub price_change_24h: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub price_change_percentage_1h: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub price_change_percentage_24h: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub price_change_percentage_7d: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub price_change_percentage_14d: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub price_change_percentage_30d: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub price_change_percentage_200d: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub price_change_percentage_1y: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub price_change_percentage_ytd: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub market_cap_change_24h: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub market_cap_change_percentage_24h: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub circulating_supply: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub total_supply: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_supply: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ath: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ath_change_percentage: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ath_date: Option<DateTime<Utc>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub atl: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub atl_change_percentage: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub atl_date: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TriggerEvaluation {
    pub trigger: PurchaseTrigger,
    pub reference_price: f64,
    pub reference_label: String,
    pub signal: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PurchaseWindowEvalResult {
    pub asset_key: String,
    pub symbol: String,
    pub current_price_usd: f64,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub trigger: Option<PurchaseTrigger>,
    pub signal: bool,
    pub reference_price: f64,
    pub reference_label: String,
    #[serde(default)]
    pub evaluations: Vec<TriggerEvaluation>,
    pub fetched_at: DateTime<Utc>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub market_data: Option<TokenMarketData>,
    #[serde(default)]
    pub data_unavailable: bool,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub error_message: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AllocationProtocol {
    pub id: String,
    pub name: String,
    #[serde(skip_serializing, default, skip_serializing_if = "Option::is_none")]
    pub logo: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AllocationChain {
    pub id: String,
    pub name: String,
    #[serde(skip_serializing, default, skip_serializing_if = "Option::is_none")]
    pub logo: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AllocationToken {
    pub symbol: String,
    pub name: String,
    pub address: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub logo: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StrategyAllocation {
    pub asset_key: String,
    pub protocol: AllocationProtocol,
    pub chain: AllocationChain,
    pub token: AllocationToken,
    pub group: String,
    pub group_type: i32,
    pub target_weight: i32,
    pub position_type: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HealthFactorTarget {
    pub asset_key: String,
    pub protocol: AllocationProtocol,
    pub chain: AllocationChain,
    pub target_health_factor: f64,
    pub critical_threshold: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StrategyDocument {
    #[serde(
        serialize_with = "serialize_uuid",
        deserialize_with = "deserialize_uuid"
    )]
    pub id: Uuid,
    pub strategy_type: StrategyType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub allocations: Option<Vec<StrategyAllocation>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub targets: Option<Vec<HealthFactorTarget>>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub purchase_window_entries: Option<Vec<BestPurchaseWindowEntry>>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub purchase_window_results: Option<Vec<PurchaseWindowEvalResult>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WalletGroupStrategies {
    #[serde(
        rename = "_id",
        serialize_with = "serialize_uuid",
        deserialize_with = "deserialize_uuid"
    )]
    pub wallet_group_id: Uuid,
    #[serde(alias = "accounts")]
    pub wallets: Vec<String>,
    pub strategies: Vec<StrategyDocument>,
    pub count: usize,
    pub key: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl WalletGroupStrategies {
    pub fn new(
        wallet_group_id: Uuid,
        wallets: Vec<String>,
        strategies: Vec<StrategyDocument>,
    ) -> Self {
        let now = Utc::now();
        let count = strategies.len();
        let key = wallet_group_id.to_string();

        Self {
            wallet_group_id,
            wallets,
            strategies,
            count,
            key,
            created_at: now,
            updated_at: now,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WalletGroupStrategiesResponse {
    pub wallet_group_id: Uuid,
    pub wallets: Vec<String>,
    pub strategies: Vec<StrategyDocument>,
    pub count: usize,
    pub key: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl From<WalletGroupStrategies> for WalletGroupStrategiesResponse {
    fn from(wgs: WalletGroupStrategies) -> Self {
        Self {
            wallet_group_id: wgs.wallet_group_id,
            wallets: wgs.wallets,
            strategies: wgs.strategies,
            count: wgs.count,
            key: wgs.key,
            created_at: wgs.created_at,
            updated_at: wgs.updated_at,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StrategyRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<Uuid>,
    pub strategy_type: StrategyType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub allocations: Option<Vec<StrategyAllocation>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub targets: Option<Vec<HealthFactorTarget>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub purchase_window_entries: Option<Vec<BestPurchaseWindowEntry>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub created_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveStrategiesRequest {
    pub wallet_group_id: String,
    pub strategies: Vec<StrategyRequest>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StrategySummary {
    pub id: Uuid,
    pub strategy_type: i32,
    pub name: Option<String>,
    pub allocations_count: usize,
    pub targets_count: usize,
    pub purchase_window_count: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveStrategiesResponse {
    pub key: String,
    pub strategies_count: usize,
    pub strategies: Vec<StrategySummary>,
    pub wallets: Vec<String>,
    pub saved_at: DateTime<Utc>,
}
